platforms/jvm/testing-jvm-infrastructure/src/main/java/org/gradle/api/internal/tasks/testing/junit/JUnitTestEventAdapter.java

```diff
@@ -210,162 +260,129 @@ private Object startParentIfNeeded(Description description) {
-        }
-        resultProcessor.started(descriptor, startEvent(parentId));
+        TestNode node = getOrRegisterNextNode(description, RegistrationMode.TEST);
+        Object parentId = startRequiredParentIfNeeded(node);
+
+        startNode(node, startEvent(parentId));
     }
 
     @Override
     public void testFailure(Failure failure) {
-        TestDescriptorInternal testInternal;
-        synchronized (lock) {
-            testInternal = executing.get(failure.getDescription());
-        }
+        addFailure(failure, null, this::reportFailure);
+    }
+
+    private void reportFailure(Object descriptorId, Throwable throwable) {
+        TestFailure failure = FAILURE_MAPPER.createFailure(throwable);
+        resultProcessor.failure(descriptorId, failure);
+    }
+
+    @Override
+    public void testAssumptionFailure(Failure failure) {
+        assumptionFailed.add(failure.getDescription());
+
+        addFailure(failure, TestResult.ResultType.SKIPPED, this::reportAssumptionFailure);
+    }
+
+    private void reportAssumptionFailure(Object descriptorId, Throwable throwable) {
+        TestFailure assumptionFailure = DefaultTestFailure.fromTestAssumptionFailure(throwable);
+        resultProcessor.failure(descriptorId, assumptionFailure);
+    }
 
-        if (testInternal != null) {
+    private void addFailure(Failure failure, TestResult.@Nullable ResultType resultType, BiConsumer<Object, Throwable> reportFailureMethod) {
+        TestNode node = getNodeOrCurrent(failure.getDescription());
+        if (node != null && !node.descriptor.isComposite() && executing.contains(node)) {
             // This is the normal path, we've just seen a test failure
             // for a test that we saw start
-            Throwable exception = failure.getException();
-            reportFailure(testInternal.getId(), exception);
+            reportFailureMethod.accept(node.descriptor.getId(), failure.getException());
         } else {
             // This can happen when, for example, a @BeforeClass or @AfterClass method fails
             // We generate an artificial start/failure/completed sequence of events
-            withPotentiallyMissingParent(className(failure.getDescription()), clock.getCurrentTime(), parentId -> {
-                TestDescriptorInternal child = nullSafeDescriptor(idGenerator.generateId(), failure.getDescription());
-                resultProcessor.started(child, startEvent(parentId));
-                Throwable exception = failure.getException();
-                reportFailure(child.getId(), exception);
-                resultProcessor.completed(child.getId(), new TestCompleteEvent(clock.getCurrentTime()));
+            withPotentiallyMissingParent(className(failure.getDescription()), clock.getCurrentTime(), parentNode -> {
+                TestNode newNode = registerNode(failure.getDescription(), RegistrationMode.TEST);
+                startNode(newNode, startEvent(parentNode.descriptor.getId()));
```

Comment:
should this be `newNode.parent.descriptor.id`? I'm not sure if parentNode will always (or should always?) be the same thing.